### ISSUE TITLE
NullPointerException in OneToManyMapping.initializeTargetPrimaryKeyFields

### ISSUE THREAD

**Issue by bleipold** [eclipse-ee4j/eclipselink#1032]
With Eclipselink 2.7.7 I get a NullPointerException when using following abstract JPA project setup (unfortunately I cannot provide the source code of the project).

1. EntityA (TableA) has a 1-n mapping to EntityB
2. EntityB (TableB) is abstract and uses single table inheritance with an integer based discriminator
3. EntityC is an implementation of EntityB wich uses a secondary table (TableC) to store some attributes

This setup leads to a NullPointerException in line 725 of method initializeTargetPrimaryKeyFields in class org.eclipse.persistence.mappings.OneToManyMapping.

I debugged a little bit and think I found the cause. The inheritance of EntityC and the secondary table leads to TableC being the default table of EntityB. In EntityAccessor.processTableAndInheritance the single table inheritance leads to TableB being not set as table of EntityC (line 1378 is skipped).

```java
if (hasInheritance() || ! rootDescriptor.usesSingleTableInheritanceStrategy()) {
    processTable();
}
```

Later on in InheritancePolicy.updateTables, since TableB is not a table of EntityC, childTables only contains the secondary table TableC and so TableC ist set as default table in line 1881.

```java
getDescriptor().setInternalDefaultTable(uniqueTables.get(uniqueTables.indexOf(childTables.get(0))));
```

Now, when the 1-n mapping from EntityA to EntityC is initialized in OneToManyMapping.initializeTargetPrimaryKeyFields the targetForeignKeyTable (TableB) dos not match the default table of EntityC (which is TableC) in line 716 and the else branch is taken.

```java
if(this.targetForeignKeyTable.equals(getReferenceDescriptor().getDefaultTable())) {
    this.targetPrimaryKeyFields = defaultTablePrimaryKeyFields;
} else {
    int sizePk = defaultTablePrimaryKeyFields.size();
    this.targetPrimaryKeyFields = new ArrayList();
    for(int i=0; i < sizePk; i++) {
        this.targetPrimaryKeyFields.add(null);
    }
    Map<DatabaseField, DatabaseField> map = getReferenceDescriptor().getAdditionalTablePrimaryKeyFields().get(this.targetForeignKeyTable);
    Iterator<Map.Entry<DatabaseField, DatabaseField>> it = map.entrySet().iterator(); // <- NullPointerException cause map is null
```

That leads to the NullPointerException in line 725 cause in line 724 getReferenceDescriptor().getAdditionalTablePrimaryKeyFields().get(targetForeignKeyTable) returns null. In my opinion, taking the else branch is wrong for this project setup.

**Comment by bleipold** [comment]
Unfortunately, I cannot provide the original entity mapping. But I will create an example next time I have time and open a bug.

No, it is no a regression on 2.7.7. Or rather, since the mapping is new, i do not know whether it would have worked on earlier releases.

**Comment by rfelcman** [comment]
Hello,
I created some test reproducer based on the bug description (see attached Maven project).
There are entities:
`com.oracle.jpa.bugtest.domain.TestEntityA` - mapped to table `TEST_TAB_A`
abstract entity `com.oracle.jpa.bugtest.domain.TestEntityB` - mapped to table `TEST_TAB_B`
there are following annotations

```
...
@Entity
@Table(name = "TEST_TAB_B")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "TYPE_ID", discriminatorType = DiscriminatorType.INTEGER)
public abstract class TestEntityB {
...
```
and "child" entities

```
...
@Entity
@DiscriminatorValue("1")
public class TestEntityB1 extends TestEntityB{
...
```

```
...
@Entity
@DiscriminatorValue("2")
public class TestEntityB2 extends TestEntityB{
...
```

Test case is functional see 'com.oracle.jpa.bugtest.TestBug'. If You want try it use _init.sql_ to create required DB tables and modify DB connection properties in _persistence.xml_ (maybe JDBC driver dependency in _pom.xml_ if You want use different DB than MySQL).
What sounds to me like collision is  **"EntityB (TableB) is abstract and uses single table inheritance with an integer based discriminator"** and **"EntityC is an implementation of EntityB wich uses a secondary table (TableC) to store some attributes"**.
If You want use single/same table for EntityB and child EntityC e.g. TableB like attached example this annotation `@Inheritance(strategy = InheritanceType.SINGLE_TABLE)` looks OK,
but if You want use two different tables (TableB, TableC) maybe in this case You have use `@Inheritance(strategy = InheritanceType.JOINED)`

But feel free to check attached test case, modify it and post it back there (GitHub) as a version 2.0.

[jpa-bug-1032-NullPointerExceptionInOneToManyMappingInitializeTargetPrimaryKeyFields.tar.gz](https://github.com/eclipse-ee4j/eclipselink/files/6189628/jpa-bug-1032-NullPointerExceptionInOneToManyMappingInitializeTargetPrimaryKeyFields.tar.gz)

**Comment by bleipold** [comment]
I have finally found time to re-investigate this issue. To be honest, I cannot reproduce it. The cause of the described behaviour is the wrong assumption, that secondary tables can be used to join additional data to an entity without the persisting part (update, delete). Eclipselink behaves as specified by the JPA-Spec unless you confusing it by doing bad things in DescriptotCustomizer (as I did and forgot to mention), to prevent it from deleting secondary table entries.
But if you have a hint for me how to join fields of additional tables to an entity without switching to projections or re-designing the data model (that's not possible), I would be very grateful.
